src/env-inl.h

```diff
@@ -983,10 +983,20 @@ inline v8::MaybeLocal<v8::Object> AllocatedBuffer::ToBuffer() {
 inline v8::Local<v8::ArrayBuffer> AllocatedBuffer::ToArrayBuffer() {
   CHECK_NOT_NULL(env_);
   uv_buf_t buf = release();
+  auto callback = [](void* data, size_t length, void* deleter_data){
+    CHECK_NOT_NULL(deleter_data);
+
+    static_cast<v8::ArrayBuffer::Allocator*>(deleter_data)
+        ->Free(data, length);
+  };
+  std::unique_ptr<v8::BackingStore> backing =
+      v8::ArrayBuffer::NewBackingStore(buf.base,
+                                       buf.len,
+                                       callback,
+                                       env_->isolate()
+                                          ->GetArrayBufferAllocator());
   return v8::ArrayBuffer::New(env_->isolate(),
-                              buf.base,
-                              buf.len,
-                              v8::ArrayBufferCreationMode::kInternalized);
+                              std::move(backing));
```

Comment:
So… This looks correct and I currently don’t have any better ideas either, but maybe to give an idea about the “big picture” here:

The idea behind `AllocatedBuffer` is to have something that gets turned into an `ArrayBuffer` in the most direct way possible. In that sense, it is basically our own version of `BackingStore` – a wrapper class for memory tracked through the current `ArrayBuffer::Allocator`. A “nicer” way to fix this would be to e.g. store the data as the `std::shared_ptr<BackingStore>` directly instead of using a `uv_buf_t`. *However*, I think that doesn’t work currently, because we’d have no way to implement e.g. `.release()` for that kind of scenario.

It might be nice if we had a way to avoid free callbacks that only run `v8::ArrayBuffer::Allocator::Free()` like this one. That would require an V8 API change, though, and I’m not sure what the best way for that would be.

Or, alternatively, it would be nice if we could implement `AllocatedBuffer` in terms of `std::shared_ptr<BackingStore>` as mentioned above; in that case, we’d have to figure out something clever for `AllocatedBuffer::release()`.

(To be clear, this isn’t a request to change the code here. It’s just me rambling and trying to give some context.)

Reply:
I'll think of something and prepare a PR for this.